### PR TITLE
[breaking] Refactoring of download subroutines

### PR DESCRIPTION
**Pull request by cmaglie**
**Please check if the PR fulfills these requirements**

- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [x] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] `UPGRADING.md` has been updated with a migration guide (for breaking changes)

**What kind of change does this PR introduce?**
The subroutines that perform download of archives and indexes have been factored.

**What is the current behavior?**
No changes in behavior

**What is the new behavior?**
No changes in behavior

**Does this PR introduce a breaking change, and is [titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?**
Some functions or methods have been moved and/or changed signature.

### CODE REVIEW

File: docs/UPGRADING.md [hunk 1]
```diff
@@ -4,6 +4,53 @@ Here you can find a list of migration guides to handle breaking changes between
...
+
+### `commands.GetDownloaderConfig` has been moved to package `github.com/arduino/arduino-cli/arduino/httpclient`
+
+All references to this function must be updated with the new import.
+
+### `commands.Download` has been removed and replaced by `github.com/arduino/arduino-cli/arduino/httpclient.DownloadFile`
+
+The old function must be replaced by the new one that is much more versatile.
+
+### `packagemanager.PackageManager.DownloadToolRelease`, `packagemanager.PackageManager.DownloadPlatformRelease`, and `resources.DownloadResource.Download` functions now require `label` and `progressCB` parameters
```
Comment (per1234) [inline docs/UPGRADING.md:23]: There was also a breaking change to their returns.

File: docs/UPGRADING.md [hunk 2]
```diff
@@ -4,6 +4,53 @@ Here you can find a list of migration guides to handle breaking changes between
 
 ## 0.22.0
 
+### package `github.com/arduino/arduino-cli/httpclient` has been moved to `github.com/arduino/arduino-cli/arduino/httpclient`
```
Comment (per1234) [inline docs/UPGRADING.md:7]: - `github.com/arduino/arduino-cli/httpclient.UserAgent` was moved to `github.com/arduino/arduino-cli/configuration.UserAgent` 
- `github.com/arduino/arduino-cli/httpclient.NetworkProxy` was moved to `github.com/arduino/arduino-cli/configuration.NetworkProxy`

**Review by per1234 — CHANGES_REQUESTED** [review]

File: configuration/network.go [hunk 3]
```diff
@@ -13,47 +13,20 @@
...
-	return &Config{
-		UserAgent: UserAgent(),
-		Proxy:     proxy,
-	}, nil
-}
-
-// UserAgent returns the user agent for the cli http client
-func UserAgent() string {
-	subComponent := configuration.Settings.GetString("network.user_agent_ext")
+// UserAgent returns the user agent for the cli (mainly used by http clients)
```
[the remark this answers is not in the document]
Comment (cmaglie) [reply configuration/network.go:27]: Uhm actually the wording `for the cli` is redundant, since this is the `configuration` package that is used in the Arduino CLI project there is no need to further specify it.

File: arduino/resources/index.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+		// Check signature...
+		if valid, _, err := security.VerifyArduinoDetachedSignature(tmpIndexPath, tmpSignaturePath); err != nil {
+			return &arduino.PermissionDeniedError{Message: tr("Error verifying signature"), Cause: err}
+		} else if !valid {
+			return &arduino.SignatureVerificationFailedError{File: res.URL.String()}
+		}
+	}
+
+	// TODO: Implement a ResourceValidator
```
Comment (umbynos) [inline arduino/resources/index.go:90]: What about this TODO? Will be done in another PR?

Comment (cmaglie) [reply arduino/resources/index.go:90]: Yes, previously we tried to parse the package_index.json with the packagemanager, but now we cannot do that anymore because we use the same subroutine to download also library_index (and potentially any other file since this function is generic enough).
